akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -30,14 +31,17 @@ final class Flow[-In, +Out, +Mat](private[stream] override val module: Module)
 
   override type Repr[+O, +M] = Flow[In @uncheckedVariance, O, M]
 
-  private[stream] def isIdentity: Boolean = this.module.isInstanceOf[Stages.Identity]
+  private[stream] def isIdentity: Boolean = this.module eq Stages.identityGraph.module
 
   def viaMat[T, Mat2, Mat3](flow: Graph[FlowShape[Out, T], Mat2])(combine: (Mat, Mat2) ⇒ Mat3): Flow[In, T, Mat3] = {
     if (this.isIdentity) {
-      val flowInstance: Flow[In, T, Mat2] = if (flow.isInstanceOf[javadsl.Flow[In, T, Mat2]])
-        flow.asInstanceOf[javadsl.Flow[In, T, Mat2]].asScala
-      else
-        flow.asInstanceOf[Flow[In, T, Mat2]]
+      val flowInstance: Flow[In, T, Mat2] =
+        if (flow.isInstanceOf[javadsl.Flow[In, T, Mat2]])
+          flow.asInstanceOf[javadsl.Flow[In, T, Mat2]].asScala
+        else if (flow.isInstanceOf[scaladsl.Flow[In, T, Mat2]])
+          flow.asInstanceOf[Flow[In, T, Mat2]]
+        else
+          Flow.wrap(flow).asInstanceOf[Flow[In, T, Mat2]]
```

Comment:
Of course the graph can be, well, a graph, not a flow even if it has that shape :). This fixes that issue.